Source/Evolve/WarpXEvolve.cpp

```diff
@@ -406,6 +420,118 @@ WarpX::OneStep_nosub (Real cur_time)
     if (warpx_py_afterEsolve) warpx_py_afterEsolve();
 }
 
+void
+WarpX::OneStep_multiJ (amrex::Real cur_time)
+{
+#ifdef WARPX_DIM_RZ
+    amrex::Abort("multi-J algorithm not implemented for RZ geometry");
+#else
+#ifdef WARPX_USE_PSATD
+    if (WarpX::maxwell_solver_id == MaxwellSolverAlgo::PSATD)
+    {
+        // Warnings:
+        // - this does not work with Galilean PSATD
+        //   (the shifts are not properly taken into account when depositing)
+        // - this does not work with PML
+
+        // Push particle from x^{n} to x^{n+1}
+        //               from p^{n-1/2} to p^{n+1/2}
+        const bool skip_deposition = true;
+        PushParticlesandDepose(cur_time, skip_deposition);
+
+        // Initialize multi-J loop:
+
+        // 1) Prepare E, B, F fields in spectral space
+        PSATDForwardTransformEB();
+        PSATDForwardTransformF();
+        PSATDForwardTransformG();
+
+        // 2) Set the averaged fields to zero
+        if (WarpX::fft_do_time_averaging) PSATDEraseAverageFields();
+
+        // 3) Deposit rho (in rho_new, since it will be moved during the loop)
+        if (WarpX::update_with_rho)
+        {
+            mypc->DepositCharge(rho_fp, -dt[0], 1);
+            // Filter, exchange boundary, and interpolate across levels
+            SyncRho();
+            // Forward FFT of rho_new
+            PSATDForwardTransformRho(1);
+        }
+
+        // 4) Deposit J if needed
+        if (WarpX::psatd_linear_in_J)
+        {
+            mypc->DepositCurrent(current_fp, dt[0], -dt[0]);
+            // Filter, exchange boundary, and interpolate across levels
+            SyncCurrent();
+            // Forward FFT of J
+            PSATDForwardTransformJ();
+        }
+
+        // Loop over multiple J deposition
+        const int n_depose = WarpX::multij_n_depose;
+        const amrex::Real sub_dt = dt[0] / static_cast<amrex::Real>(n_depose);
+        const int n_loop = (WarpX::fft_do_time_averaging) ? 2*n_depose : n_depose;
+        for (int i_depose = 0; i_depose < n_loop; i_depose++)
+        {
+            // Move previously deposited rho and J
+            PSATDMoveRhoNewToRhoOld();
+            if (WarpX::psatd_linear_in_J) PSATDMoveJNewToJOld();
+
+            // Deposit the new J
+            const amrex::Real t_depose = (WarpX::psatd_linear_in_J) ?
+                (i_depose-n_depose+1)*sub_dt : (i_depose-n_depose+0.5)*sub_dt;
+            mypc->DepositCurrent(current_fp, dt[0], t_depose);
+            // Filter, exchange boundary, and interpolate across levels
+            SyncCurrent();
+            // Forward FFT of J
+            PSATDForwardTransformJ();
+
+            // Deposit the rho_new
+            if (WarpX::update_with_rho)
+            {
+                mypc->DepositCharge(rho_fp, (i_depose-n_depose+1)*sub_dt, 1);
+                // Filter, exchange boundary, and interpolate across levels
+                SyncRho();
```

Comment:
This is probably inefficient because it performs filtering, boundary exchange and interpolate on the 2 components of rho in real space (rho_new and rho_old)